File: solutions/javascript/robot-simulator/2/robot-simulator.js
/**
 * Class representing an error for invalid input.
 * @extends Error
 */
export class InvalidInputError extends Error {
  /**
   * Create an InvalidInputError.
   * @param {string} [message='Invalid Input'] - The error message.
   */
  constructor(message) {
    super();
    this.message = message || 'Invalid Input';
  }
}

/**
 * Class representing a Robot.
 */
export class Robot {
  /**
   * Get the instructions for the robot.
   * @param {string} s - The string of instructions.
   * @returns {Array<string>} The array of instruction methods.
   * @throws {InvalidInputError} If the instruction character is invalid.
   */
  static instructions(s) {
    return [...s].map((character) => {
      switch (character) {
        case 'L':
          return 'turnLeft';
        case 'R':
          return 'turnRight';
        case 'A':
          return 'advance';
        default:
          throw new InvalidInputError(
            `${character} is not a valid instruction character.`,
          );
      }
    });
  }

  /**
   * Create a Robot.
   */
  constructor() {
    this.coordinates = [0, 0];
    this.bearing = 'north';
  }

  /**
   * Set the direction of the robot.
   * @param {string} next - The next direction.
   * @throws {InvalidInputError} If the direction is invalid.
   */
  set direction(next) {
    const validDirections = ['north', 'south', 'east', 'west'];
    if (!validDirections.includes(next)) {
      throw new InvalidInputError('Invalid Robot Bearing');
    }

    this.bearing = next;
  }

  /**
   * Advance the robot in the current direction.
   */
  advance() {
    switch (this.bearing) {
      case 'north':
        this.coordinates[1] += 1;
        break;
      case 'south':
        this.coordinates[1] -= 1;
        break;
      case 'east':
        this.coordinates[0] += 1;
        break;
      case 'west':
        this.coordinates[0] -= 1;
        break;
      default:
        console.warn(`Unknown bearing: ${this.bearing}`);
        break;
    }
  }

  /**
   * Turn the robot to the left.
   */
  turnLeft() {
    switch (this.bearing) {
      case 'north':
        this.direction = 'west';
        break;
      case 'south':
        this.direction = 'east';
        break;
      case 'east':
        this.direction = 'north';
        break;
      case 'west':
        this.direction = 'south';
        break;
      default:
        console.warn(`Unknown bearing: ${this.bearing}`);
        break;
    }
  }

  /**
   * Turn the robot to the right.
   */
  turnRight() {
    switch (this.bearing) {
      case 'north':
        this.direction = 'east';
        break;
      case 'south':
        this.direction = 'west';
        break;
      case 'east':
        this.direction = 'south';
        break;
      case 'west':
        this.direction = 'north';
        break;
      default:
        console.warn(`Unknown bearing: ${this.bearing}`);
        break;
    }
  }

  /**
   * Place the robot at specific coordinates and direction.
   * @param {Object} args - The placement arguments.
   * @param {number} args.x - The x-coordinate.
   * @param {number} args.y - The y-coordinate.
   * @param {string} args.direction - The direction.
   */
  place(args) {
    this.coordinates = [args.x, args.y];
    this.direction = args.direction;
  }

  /**
   * Evaluate a string of instructions.
   * @param {string} s - The string of instructions.
   */
  evaluate(s) {
    Robot.instructions(s).forEach((instruction) => {
      this[instruction]();
    });
  }
}

File: solutions/python/robot-simulator/2/robot_simulator.py
"""
This module defines a robot that can be moved around on a grid.
"""

EAST = 0
NORTH = 1
WEST = 2
SOUTH = 3


class Robot:
    """
    A class to represent a robot with position and direction on a grid.
    """

    def __init__(
        self,
        direction: int = NORTH,
        x_coordinate: int = 0,
        y_coordinate: int = 0,
    ) -> None:
        """
        Initializes the robot with a given direction and coordinates.

        :param direction: Initial direction the robot is facing (default is NORTH).
        :param x_coordinate: Initial x-axis coordinate of the robot (default is 0).
        :param y_coordinate: Initial y-axis coordinate of the robot (default is 0).
        """
        self.direction: int = direction
        self.x_coordinate: int = x_coordinate
        self.y_coordinate: int = y_coordinate
        self.coordinates: tuple[int, int] = (
            self.x_coordinate,
            self.y_coordinate,
        )

    def move(self, instructions: str = "") -> None:
        """
        Moves the robot on the grid based on the given instructions.

        :param instructions: A string of instructions where 'R' turns the robot right,
                             'L' turns the robot left, and 'A' moves the robot forward in
                             its current direction.
        :raises ValueError: If an invalid instruction is provided.
        """
        moves: dict[int, tuple[int, int]] = {
            EAST: (1, 0),
            NORTH: (0, 1),
            WEST: (-1, 0),
            SOUTH: (0, -1),
        }
        for instruction in instructions:
            if instruction == "R":
                self.direction = (self.direction - 1) % 4
            elif instruction == "L":
                self.direction = (self.direction + 1) % 4
            elif instruction == "A":
                delta_x, delta_y = moves[self.direction]
                self.x_coordinate += delta_x
                self.y_coordinate += delta_y
                self.coordinates = self.x_coordinate, self.y_coordinate
            else:
                raise ValueError("Wrong Instruction Given To Robot!")
